compiler/backend/src/org/jetbrains/kotlin/codegen/ClassBuilderFactories.java

```diff
@@ -135,9 +135,13 @@ protected String getCommonSuperClass(@NotNull String type1, @NotNull String type
             // This method is needed to generate StackFrameMap: bytecode metadata for JVM verification. For bytecode version 50.0 (JDK 6)
             // these maps can be invalid: in this case, JVM would generate them itself (potentially slowing class loading),
             // for bytecode 51.0+ (JDK 7+) JVM would crash with VerifyError.
-            // It seems that for bytecode emitted by Kotlin compiler, it is safe to return "Object" here, because there will
-            // be "checkcast" generated before making a call, anyway.
 
+            if (AsmTypeInfo.INSTANCE.isSubType(type1, type2)) {
+                return type2;
+            }
+            if (AsmTypeInfo.INSTANCE.isSubType(type2, type1)) {
+                return type1;
+            }
             return "java/lang/Object";
         }
     }
```

Comment:
This time the problem is related to [StackMapTable](https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.7.4). Since `getCommonSuperClass` directly returns `java/lang/Object`, `CHECKCAST` is essential in the above failed code (I have added a similar test case, see `kt11700_stackMapTable.kt` ).

I don't know to use the compiler generated by `gradlew dist` to compile this repo itself, considering that the coverage of tests may still not be comprehensive enough, I should done that before pushing my code, sorry about that.